### PR TITLE
[RISCV] Add missing hasPostISelHook = 1 to vector pseudos that might read FRM.

### PR DESCRIPTION
**Pull request by topperc**
We need an implicit FRM read operand anytime the rounding mode is dynamic. The post isel hook is responsible for this when isel creates an instruction with dynamic rounding mode.

Add a MachineVerifier check to verify the operand is present.

### CODE REVIEW

**Review by wangpc-pp — COMMENTED** [review]
I don't know if I understand correctly, we don't add `frm` operand for some cases before?

**Comment by topperc** [comment]
> I don't know if I understand correctly, we don't add `frm` operand for some cases before?

That's correct. I think InsertReadWriteCSR would add it in the tests I looked at, but that pass really isn't supposed to anything for DYN instructions.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 1]
```diff
@@ -2617,6 +2617,13 @@ bool RISCVInstrInfo::verifyInstruction(const MachineInstr &MI,
     }
   }
 
+  if (int Idx = RISCVII::getFRMOpNum(Desc);
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2620]: This is checking the pseudo form, is there any way that we can add an additional assert which checks the pseudo against the underlying instruction?  The underlying instruction needs to have the use of FRM too doesn't it?

(Possible follow up)

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td [hunk 2]
```diff
@@ -6504,7 +6504,7 @@ defm PseudoVFWMACCBF16  : VPseudoVWMAC_VV_VF_BF_RM;
 //===----------------------------------------------------------------------===//
 // 13.8. Vector Floating-Point Square-Root Instruction
 //===----------------------------------------------------------------------===//
-let mayRaiseFPException = true, hasSideEffects = 0 in
+let mayRaiseFPException = true, hasSideEffects = 0, hasPostISelHook = 1 in
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:6507]: It doesn't look like we have support for this today, but maybe we should add a target hook for this instead?  Having the post ISEL hook driven by tablgen (as opposed to a query on the underlying instruction), seems error prone.  Another idea might be a target specific consistency check before emission so at least we know at build time that pseudos and instructions are in sync?

**Review by preames — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td (hunk 2 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:6507]: I was planning to move the hasPostISelHook=1 into the base class where has HasRoundMode is set. The flag should be set anytime HasRoundMode==1 && UsesVXRM=0.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 1 above)
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2620]: Wouldn't a disassembly use case be wrong if we didn't have the implicit use?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2620]: It doesn't affect the printing of the instruction so the main use of disassembly wouldn't notice.

**Pull request merged**